crates/wysiwyg/src/dom/action_list.rs

```diff
@@ -0,0 +1,316 @@
+        })
+    }
+
+    pub fn remove_node(handle: DomHandle) -> Self {
+        DomAction::Remove(RemoveNodeAction { handle })
+    }
+
+    pub fn move_node(from_handle: DomHandle, to_handle: DomHandle) -> Self {
+        DomAction::Move(MoveNodeAction {
+            from_handle,
+            to_handle,
+        })
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct AddNodeAction<S: UnicodeString> {
+    pub parent_handle: DomHandle,
+    pub index: usize,
+    pub node: DomNode<S>,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct RemoveNodeAction {
+    pub handle: DomHandle,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct MoveNodeAction {
+    pub from_handle: DomHandle,
+    pub to_handle: DomHandle,
+}
+
+/// Ordered list of DomActions to be performed. It contains helper methods to access different kind
+/// of actions.
+#[derive(Clone, PartialEq, Debug)]
+pub struct DomActionList<S: UnicodeString> {
+    actions: Vec<DomAction<S>>,
+}
+
+impl<S: UnicodeString> DomActionList<S> {
+    #[allow(dead_code)]
+    pub fn new(actions: Vec<DomAction<S>>) -> Self {
+        Self { actions }
+    }
+
+    #[allow(dead_code)]
+    pub fn actions(&self) -> &Vec<DomAction<S>> {
+        &self.actions
+    }
+
+    #[allow(dead_code)]
+    pub fn additions(&self) -> Vec<&AddNodeAction<S>> {
+        self.actions
+            .iter()
+            .filter_map(|a| match a {
+                DomAction::Add(a) => Some(a),
+                _ => None,
+            })
+            .collect()
+    }
+
+    #[allow(dead_code)]
+    pub fn removals(&self) -> Vec<&RemoveNodeAction> {
+        self.actions
+            .iter()
+            .filter_map(|a| match a {
+                DomAction::Remove(a) => Some(a),
+                _ => None,
+            })
+            .collect()
+    }
+
+    #[allow(dead_code)]
+    pub fn moves(&self) -> Vec<&MoveNodeAction> {
+        self.actions
+            .iter()
+            .filter_map(|a| match a {
+                DomAction::Move(a) => Some(a),
+                _ => None,
+            })
+            .collect()
+    }
+
+    /// To avoid unnecessary copies, this function returns the [actions] and marks the DomActionList
+    /// as moved.
+    pub fn grouped(
+        self,
+    ) -> (
+        Vec<AddNodeAction<S>>,
+        Vec<RemoveNodeAction>,
+        Vec<MoveNodeAction>,
+    ) {
+        let mut to_add = Vec::new();
+        let mut to_remove = Vec::new();
+        let mut to_move = Vec::new();
+
+        for action in self.actions {
+            match action {
+                DomAction::Add(a) => to_add.push(a),
+                DomAction::Remove(a) => to_remove.push(a),
+                DomAction::Move(a) => to_move.push(a),
+            }
+        }
+        (to_add, to_remove, to_move)
+    }
+
+    pub fn push(&mut self, action: DomAction<S>) {
+        self.actions.push(action);
+    }
+
+    #[allow(dead_code)]
+    pub fn replace_actions(&mut self, new_actions: Vec<DomAction<S>>) {
```

Comment:
Why do we need this? Can't we just create a new DomActionList?

Reply:
If you're passing a reference to the current `DomActionList` you might need to replace its contents while keeping the same reference.